pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -647,6 +660,13 @@ func (cm *containerManagerImpl) GetPluginRegistrationHandler() cache.PluginHandl
 // TODO: move the GetResources logic to PodContainerManager.
 func (cm *containerManagerImpl) GetResources(pod *v1.Pod, container *v1.Container) (*kubecontainer.RunContainerOptions, error) {
 	opts := &kubecontainer.RunContainerOptions{}
+	if cm.draManager != nil {
+		resOpts, err := cm.draManager.PrepareResources(pod, container)
```

Comment:
@pohly  @klueska 
The PrepareResources is called after the CNI is invoked. This is a problem for network devices that want to use DRA framework. 

In our case we have complex netdevice with several of other devices that we want to map to container and we rely on DRA to create that network device. The CNI call need the netdevice to be ready so we can move it to network namespace.  The CDI will mount the other related devices to the container. 

This worked well with device plugin.

I think the difference between device plugin and DRA is that in device plugin kubelet allocated the device at  podadmit [1] and in the container-manager just update the container config ,but for DRA all is done in container_manager. 
 Is is possible to call PrepareResources be before Pod generation?  another option is to split PrepareResources  to 2 Api calls PrepareResources  on admit and GetCDIOptions in the container manager (same as device plugin does L670-L672)

What do you think? 


[1] - https://github.com/kubernetes/kubernetes/blob/127f33f63d118d8d61bebaba2a240c60f71c824a/pkg/kubelet/cm/container_manager_linux.go#L692


around [1] https://github.com/kubernetes/kubernetes/blob/127f33f63d118d8d61bebaba2a240c60f71c824a/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L757

Reply:
The initial implementation for `PrepareResources()` was done during the pod admission phase (just as we do with the device manager) and a conscious decision was made to move it to pod creation. I'll let @bart0sh elaborate on why we moved it and what the challenges having it during pod admit were.

Reply:
We moved it because we wanted NodePrepareResource call to be repeated if it doesn't succeed. It was not possible to do on admission. However, we still can call it on admission phase and ignore its errors and then repeat on container creation if it errored on admission. @moshe010 would this work for your use case?

Reply:
I think it will be the same problem that we won't have retry if the creation of the netdevice failed because as you mentioned that  NodePrepareResource won't be repeated if it doesn't succeed on admit, So the CNI will failed and we will be in a loop Pod creation. 

The cni invoke is here [1] maybe we can add another call for NodePrepareResource  and the  SyncPod will handle the retry. What do you think?
[1] - https://github.com/kubernetes/kubernetes/blob/5539a5b80fe45cc04df4f53bd1047cd6d998bcf1/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L761

Reply:
Are you suggesting to have one call that gets called twice, or two different calls with different semantic?

Perhaps the ultimate problem is that kubelet treats errors during admission as fatal? It calls out to external components, right? Something can always go wrong only temporarily, so perhaps it should support retry unless the error clearly indicates that it is a permanent condition?

Reply:
I think we can do it with them same call twice and the DRA plugin(s)s should know to handle already created devices. 

I agree that if it possibly to change admission to support retry it better.

Reply:
@moshe010 > The cni invoke is here [1] maybe we can add another call for NodePrepareResource and the SyncPod will handle the retry. What do you think?

That's a good idea. I can try to move NodePrepareResource call to that place and see if that works.

Reply:
@moshe010 I've moved PrepareResource call to the place you've pointed out. All our tests pass.
PTAL: https://github.com/pohly/kubernetes/pull/43

Reply:
This change won't make it in for today's code freeze (assuming we make it). We will consider it for inclusion after the code freeze.

Reply:
@moshe010 
> I think the difference between device plugin and DRA is that in device plugin kubelet allocated the device at podadmit [1] and in the container-manager just update the container config ,but for DRA all is done in container_manager.

The main difference between device plugin framework and DRA is that DRA was designed with an assumption that resource allocation should happen **before** the workload is even scheduled to the node.

Node API calls (_NodePrepareResources_ and _NodeUnprepareResources_) are not expected to allocate resources, unlike device plugin's _Allocate_ call.
Here is a quote from the KEP describing NodePrepareResource API https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation#NodePrepareResource
```
This operation SHALL do as little work as possible as it’s called after a pod is scheduled to a node.
All potentially failing operations SHALL be done during allocation phase.
```

Is there a possibility to follow this design for network devices?

PS: This doesn't mean that we can't move the call before CNI init. I'm just wondering if there _always_ be a need for real device allocation on the node for network devices.

Reply:
In my case the devices (e.g sub functions) are dynamically created. In the  allocation phase I will just allocate free sub functions index and in NodePrepareResource I will create and setup the device .  I don't understand the "operation SHALL do as little work as possible" in my understanding NodePrepareResource  should be the phase to configure the device. (in my case I need to create and configure sub functions) Also it not clear "All potentially failing operations SHALL be done during allocation" in the NodePrepareResource  we configuring the hardware device so anyway we can expect failures. 

Is my understanding is wrong?

Reply:
Your understanding is correct. However, NodePrepareResource should never fail because the node ran out of resources. That is a situation that the resource driver should avoid during the allocation phase, before a pod gets scheduled onto the node.